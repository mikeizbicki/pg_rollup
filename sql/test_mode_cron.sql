create or replace language plpython3u;
create extension if not exists hll;
create extension if not exists pg_rollup;
create extension if not exists pg_cron;

create temporary table test_cron (
    id serial primary key,
    name text,
    num int
);

insert into test_cron (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

UPDATE pg_rollup_settings SET value='cron' WHERE name='default_mode';

select create_rollup(
    'test_cron',
    'test_cron_rollup1',
    wheres => 'name',
    key => 'id'
);

select create_rollup(
    'test_cron',
    'test_cron_rollup2',
    wheres => 'name,num',
    key => 'id'
);

select create_rollup(
    'test_cron',
    'test_cron_rollup3',
    wheres => 'name',
    rollups => 'hll(num)',
    key => 'id',
    mode => 'trigger'
);

select create_rollup(
    'test_cron',
    'test_cron_rollup4',
    rollups => 'hll(name),hll(num)',
    key => 'id',
    mode => 'cron'
);

insert into test_cron (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

-- cron jobs do not run in the environment created by the make installcheck command;
-- therefore, the rollup commands will not get executed and the tables will be out of date;
-- to test_cron the cron mode, therefore, we will inspect the output of the cron job list
-- and verify that all jobs that should be added have been added
select * from cron.job;

UPDATE pg_rollup_settings SET value='trigger' WHERE name='default_mode';
