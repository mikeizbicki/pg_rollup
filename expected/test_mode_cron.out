create or replace language plpython3u;
create extension if not exists hll;
NOTICE:  extension "hll" already exists, skipping
create extension if not exists pg_rollup;
NOTICE:  extension "pg_rollup" already exists, skipping
create extension if not exists pg_cron;
create table test (
    id serial primary key,
    name text,
    num int
);
insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
UPDATE pg_rollup_settings SET value='cron' WHERE name='default_mode';
select create_rollup(
    'test',
    'test_rollup1',
    wheres => 'name',
    key => 'id'
);
 create_rollup 
---------------
 
(1 row)

select create_rollup(
    'test',
    'test_rollup2',
    wheres => 'name,num',
    key => 'id'
);
 create_rollup 
---------------
 
(1 row)

select create_rollup(
    'test',
    'test_rollup3',
    wheres => 'name',
    distincts => 'num',
    key => 'id',
    mode => 'trigger'
);
 create_rollup 
---------------
 
(1 row)

select create_rollup(
    'test',
    'test_rollup4',
    distincts => 'name,num',
    key => 'id',
    mode => 'cron'
);
 create_rollup 
---------------
 
(1 row)

insert into test (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
-- cron jobs do not run in the environment created by the make installcheck command;
-- therefore, the rollup commands will not get executed and the tables will be out of date;
-- to test the cron mode, therefore, we will inspect the output of the cron job list
-- and verify that all jobs that should be added have been added
select * from cron.job;
 jobid | schedule  |                       command                       | nodename  | nodeport | database | username | active |        jobname         
-------+-----------+-----------------------------------------------------+-----------+----------+----------+----------+--------+------------------------
     1 | * * * * * | SELECT do_rollup('test_rollup1',delay_seconds=>0);  | localhost |     5432 | root     | root     | t      | pg_rollup.test_rollup1
     2 | * * * * * | SELECT do_rollup('test_rollup2',delay_seconds=>13); | localhost |     5432 | root     | root     | t      | pg_rollup.test_rollup2
     3 | * * * * * | SELECT do_rollup('test_rollup4',delay_seconds=>26); | localhost |     5432 | root     | root     | t      | pg_rollup.test_rollup4
(3 rows)

